lib/TokenManager.ts

```diff
@@ -40,6 +40,14 @@ function emitExpired(tokenMgmtRef, key, token) {
   tokenMgmtRef.emitter.emit('expired', key, token);
 }
 
+function emitRenewed(tokenMgmtRef, key, freshToken, oldToken) {
+  tokenMgmtRef.emitter.emit('renewed', key, freshToken, oldToken);
+}
+
+function emitAdded(tokenMgmtRef, key, token) {
+  tokenMgmtRef.emitter.emit('added', key, token);
+}
+
 function emitRemoved(tokenMgmtRef, key) {
   tokenMgmtRef.emitter.emit('removed', key);
```

Comment:
Do we have to worry about the removed event emitting late?  Should we emit the removed token as well as the key?

Reply:
I think it should be fine. This event only emit after token is removed, then we only need the key to update the synced in-memory copy.

Reply:
It would probably not hurt to emit the removed token, but I'm not sure what anyone would do with it. The key makes sense to remove from any synced copies, as @shuowu  said. If there is a use case for doing something with a removed token, it would be fine to emit it here.

Reply:
My main concern is if the removed event gets handled after the token is already replaced.  They wouldn't have a way to know if the event applies to the current token.

Reply:
I think that is a good case. It's possible that some time may pass between the 'removed' event and when the handler decides to do something. In that time, the token at that key could have been replaced by, say, a successful `getWithoutPrompt` call. So they could compare the token with the one in storage. One problem is that the token instance will always be different since it is deserialized from storage. We could:
- try to save a copy of token in memory and avoid replacing it if it matches object in storage, then user can compare object references to detect if tokens are different. This would be at the layer above the storage provider. It would be expected storage providers would always return a fresh object for a key.
- provide a helper method (similar to `hasExpired` which will do the token compare. Something like `isEqual`, then we would look at a certain string field to equal (probably token.idToken or token.accessToken)

Reply:
@aarongranick-okta By the `handler`, do you mean the `isAuthenticated` handler? I still don't think it can be the case as long as we can manage all processes properly in the promiseQ. I updated the implementation here: https://github.com/okta/okta-auth-js/pull/453/files#diff-75082c639aefcc47b3b8df57c8870867R331

Reply:
By `handler`, I mean the function which is called when the `removed` event is fired. I think it would be a good idea to pass the token which was removed as the second argument so we know **exactly** which token was removed and it can be compared to any other token. But this raises the question of how to compare tokens (by object reference or via util function/string compare).

Reply:
Added token for `removed` event. Right now I don't see the need to compare the tokens, but there is no hurt to just expose it from the event.